src/main/java/org/gaul/s3proxy/junit/S3ProxyExtension.java

```diff
@@ -0,0 +1,189 @@
+/*
+ * Copyright 2014-2020 Andrew Gaul <andrew@gaul.org>
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gaul.s3proxy.junit;
+
+import java.io.File;
+import java.io.IOException;
+import java.net.URI;
+import java.nio.file.Files;
+import java.util.Properties;
+
+import org.apache.commons.io.FileUtils;
+import org.eclipse.jetty.util.component.AbstractLifeCycle;
+import org.gaul.s3proxy.AuthenticationType;
+import org.gaul.s3proxy.S3Proxy;
+import org.jclouds.ContextBuilder;
+import org.jclouds.blobstore.BlobStore;
+import org.jclouds.blobstore.BlobStoreContext;
+import org.jclouds.blobstore.domain.StorageMetadata;
+import org.junit.jupiter.api.extension.AfterAllCallback;
+import org.junit.jupiter.api.extension.BeforeAllCallback;
+import org.junit.jupiter.api.extension.ExtensionContext;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+public final class S3ProxyExtension
```

Comment:
There is a *lot* of repeated code between this and `S3ProxyRule`. I have worked on migrating quite a few projects to support JUnit4 Rules and JUnit5 extensions and this is how I have approached it:

* Move all the common, non-JUnit specific code out from the existing Rule into a new "Core" class.
* Have the Rule call out to this "Core" class from the various JUnit lifecycle methods.
* All existing unit tests and usages of the Rule shouldn't need to change.
* Implement the Extension in a similar manner to call out to the "Core" from the various JUnit5 lifecyle methods it implements.
* Add a test for the Extension along the same lines as the Rule test.
* Decide on which JUnit jars you want to mark with "compile", "test" or "provided" scopes in Maven depending how the project is used downstream and whether you want users to inherit JUnit dependencies transitively or whether you want them to explicitly pick and choose the JUnit versions they want.
* Update documentation describing how to use JUnit5 extension alongside the JUnit4 rule so users can use either.

Reply:
I'm totally ok with the code duplication comment.

I'm struggling with it because the JUnit 4 rule extends some Junit class and the JUnit 5 extension implements some JUnit interface.

I've already added a test for the extension (4th bullet)

Reply:
I think the only way you can refactor out the common code is using composition rather than inheritance since the JUnit4 Rule already extends another class like you said. Here is an example that might be helpful - [BeejuJunitExtension](https://github.com/HotelsDotCom/beeju/blob/main/src/main/java/com/hotels/beeju/extensions/BeejuJUnitExtension.java) and [BeejuJUnitRule](https://github.com/HotelsDotCom/beeju/blob/main/src/main/java/com/hotels/beeju/BeejuJUnitRule.java) both delegate most of the functionality to [BeejuCore](https://github.com/HotelsDotCom/beeju/blob/main/src/main/java/com/hotels/beeju/core/BeejuCore.java) (which in turn doesn't actually know anything about JUnit). The extension and the rule keep extending/implementing the same JUnit classes and interfaces, you just move the code which isn't dependent on a specific version of JUnit out into the core.

Reply:
Yes that's the way it shall be done.
Will do as soon as I can.

Reply:
I've done the factorization. I've kept the Builder in each class to avoid breaking existing JUnit 4 test classes.